src/main/java/io/vertx/core/streams/impl/OutboundWriteQueue.java

```diff
@@ -0,0 +1,361 @@
+ * expects another thread or another condition to be met to make progress. Emitting methods can return a signal indicating that
+ * the queue should be drained to make progress: {@link #DRAIN_REQUIRED_MASK} signals the queue contains element that
+ * shall be drained.
+ *
+ * <h3>Draining elements</h3>
+ *
+ * The {@link #drain} method tries to remove elements from the queue until the consumer refuses them or the queue becomes empty.
+ * When a method returns a {@link #DRAIN_REQUIRED_MASK} signal, the {@link #drain()} should be called by the consumer thread
+ *
+ * <ul>
+ *   <li>A consumer method ({@link #add}, {@link #drain}) emits this signal when the {@link #consumer} refuses an element,
+ *   therefore {@link #drain()} should be called when the consumer can accept element again</li>
+ *   <li>A producer method ({@link #submit}) emits this signal when it acquired the ownership of the queue to {@link #drain} the
+ *   queue.</li>
+ * </ul>
+ *
+ * <h3>Back-pressure</h3>
+ *
+ * <p>Producers emission flow cannot reliably be controlled by the consumer back-pressure probe since producers
+ * emissions are transferred to the consumer thread: the consumer back-pressure controller does not take in account the
+ * inflight elements between producers and consumer. This queue is designed to provide reliable signals to control
+ * producers emission based on the consumer back-pressure probe and the number of inflight elements between
+ * producers and the consumer.</p>
+ *
+ * The queue maintains an internal queue of elements initially empty and can be filled when
+ * <ul>
+ *   <li>producer threads {@link #submit)} to the queue above the {@link #highWaterMark}</li>
+ *   <li>the {@link #consumer} refuses an element</li>
+ * </ul>
+ *
+ * <p>When the internal queue grows above the {@link #highWaterMark}, the queue is considered as {@code unwritable}.
+ * {@link #add}/{@link #submit} methods return {@link #QUEUE_UNWRITABLE_MASK} to signal producers should stop emitting.</p>
+ *
+ * <p>After a drain if the internal queue has shrunk under {@link #lowWaterMark}, the queue is considered as {@code writable}.
+ * {@link #add}/{@link #drain} methods return {@link #QUEUE_WRITABLE_MASK} to signal producers should start emitting. Note
+ * the consumer thread handle this signal and should transmit it to the producers.</p>
+ *
+ * <p>When {@link #QUEUE_WRITABLE_MASK} is signalled, the number of {@link #QUEUE_UNWRITABLE_MASK} signals emitted is encoded
+ * in the flags. This number allows the producer flow controller to correctly account the producer writability:
+ * {@link #QUEUE_WRITABLE_MASK}/{@link #QUEUE_UNWRITABLE_MASK} signals are observed by different threads, therefore a good
+ * practice to compute the queue writability is to increment/decrement an atomic counter for each signal received.</p>
+ */
+public class OutboundWriteQueue<E> {
```

Comment:
A point about the bitmasks. I am a big fan of them, but not to expose them externally because they reveal of things are handled internally and bound you to never hange them.
If we can expose enums or other things that decide which states leaks to become part of the API would be better, IMO. 
At the same time, I understand that if we need to return set of states we have no other options, to save allocations...

Reply:
one of the mask returns an positive value encoded